src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -10,45 +10,54 @@ namespace Microsoft.CodeAnalysis.Text
 {
     internal sealed class ChangedText : SourceText
     {
-        private readonly SourceText _oldText;
         private readonly SourceText _newText;
+
+        // store old text weakly so we don't form unwanted chains of old texts (especially chains of ChangedTexts)
+        // It is only used to identify the old text in GetChangeRanges which only returns the changes if old text matches identity.
+        private readonly WeakReference<SourceText> _weakOldText;
         private readonly ImmutableArray<TextChangeRange> _changes;
 
-        public ChangedText(SourceText oldText, ImmutableArray<TextChangeRange> changeRanges, ImmutableArray<SourceText> segments)
+        public ChangedText(SourceText oldText, SourceText newText, ImmutableArray<TextChangeRange> changeRanges)
             : base(checksumAlgorithm: oldText.ChecksumAlgorithm)
         {
+            Debug.Assert(newText != null);
             Debug.Assert(oldText != null);
+            Debug.Assert(oldText != newText);
             Debug.Assert(!changeRanges.IsDefault);
-            Debug.Assert(!segments.IsDefault);
 
-            _oldText = oldText;
-            _newText = segments.IsEmpty ? new StringText("", oldText.Encoding, checksumAlgorithm: oldText.ChecksumAlgorithm) : (SourceText)new CompositeText(segments);
+            _newText = newText;
+            _weakOldText = new WeakReference<SourceText>(oldText);
             _changes = changeRanges;
         }
 
         public override Encoding Encoding
         {
-            get { return _oldText.Encoding; }
+            get { return _newText.Encoding; }
         }
 
-        public SourceText OldText
+        public IEnumerable<TextChangeRange> Changes
         {
-            get { return _oldText; }
+            get { return _changes; }
         }
 
-        public SourceText NewText
+        public override int Length
         {
-            get { return _newText; }
+            get { return _newText.Length; }
         }
 
-        public IEnumerable<TextChangeRange> Changes
+        internal override int Size
```

Comment:
Can we rename this to "UnderlyingStorageSize" or something like that? The tests gave me some head scratching where both are being asserted and they're different.